### DISCUSSION TITLE
wrong return

### DISCUSSION THREAD

**Discussion by Larbo53** [jsvine/pdfplumber#596] [Ask for help with specific PDFs]
Hello,

I extract a zone with the command :
invoice=doc.pages[i].crop((row[2],row[3],row[4],row[5]),relative = True).extract_text()

the returned value is : 9161F42A87C
while the display in the pdf is : 91614287.
How can I find the right value? Is it a problem with the fonts or the pdf editor?

Thanks for your feedback.

Best regards.
[Valhrona_Facture n°91614287.pdf](https://github.com/jsvine/pdfplumber/files/7995861/Valhrona_Facture.n.91614287.pdf)

**Comment by jsvine** [comment]
What is `row` here? Or, more generally, can you share a full code snippet that reproduces your issue?

**Reply by Larbo53** [reply]
Extract of the code 👍 

sql="SELECT Code_fournisseurs,Nom_colonne,X0,Y0,X1,Y1 FROM Schema_recup_Data\
 WHERE Nom_colonne='Num_facture' AND Code_fournisseurs='VALRHONA'"  # récupérer les coordonnées du Num de facture
curs2.execute(sql)
for row in curs2:
    for i in range(0,len(doc.pages)

**Reply by Larbo53** [reply]
All Code 

import pdfplumber 
import sqlite3
from pathlib import Path

conn=sqlite3.connect('/Users/yves/Recup-Z-interface-compta/parametrage_facture/BD_parametrage_facture_fournisseurs.db') # Base de données fournisseurs
curs2=conn.cursor()

path='/Users/yves/Documents/Factures/'
Identifiant=path+'Valhrona_Facture n°91614287.pdf'
Fichiers_Identifiant={}

doc = pdfplumber.open(Identifiant)
Pages_Factures={}

sql="SELECT Code_fournisseurs,Nom_colonne,X0,Y0,X1,Y1 FROM Schema_recup_Data\
 WHERE Nom_colonne='Num_facture' AND Code_fournisseurs='VALRHONA'"  # récupérer les coordonnées du Num de facture
curs2.execute(sql)
for row in curs2:
    for i in range(0,len(doc.pages)):#vérifier que c'est un num conforme c'est à dire qu'il contient au moins 2 chiffres
        invoice=doc.pages[i].crop((row[2],row[3],row[4],row[5]),relative = True).extract_text()
        if int(invoice[-2])==False:
            continue
        else:
            print(i)
            Pages_Factures[i+1]=invoice

**Reply by jsvine** [reply]
This isn't exactly reproducible, since the sqlite3 database isn't attached. But it looks like the `FAC` from the `FACTURE` above the number is merging into the number itself. That's strange, though, because I don't seem to be able to reproduce the problem — when the value appears correct when I run this simple example:

```python
import pdfplumber
pdf = pdfplumber.open("Valhrona_Facture.n.91614287.pdf")
page = pdf.pages[0]
print(page.crop((400, 50, 460, 70)).extract_text())
```

Result: `91614287`

What version of `pdfplumber` are you using?

---

Also, you can provide your code an easier-to-read format by enclosing it in tick marks:

````
```python
a = True
```
````

... displays:

```python
a = True
```

**Reply by Larbo53** [reply]
your answer 👍 👍 
This isn't exactly reproducible, since the sqlite3 database isn't attached. But it looks like the FAC from the FACTURE above the number is merging into the number itself. That's strange, though, because I don't seem to be able to reproduce the problem — when the value appears correct when I run this simple example:

import pdfplumber
pdf = pdfplumber.open("Valhrona_Facture.n.91614287.pdf")
page = pdf.pages[0]
print(page.crop((400, 50, 460, 70)).extract_text())
Result: 91614287

What version of pdfplumber are you using?

Also, you can provide your code an easier-to-read format by enclosing it in tick marks:

```python
a = True
```
... displays:

a = True


Thanks a lot.
1 - The row returns these values.
How to knows the pdfplumber’s version?

2 - I don't understand your second remark :
"Also, you can provide your code an easier-to-read format by enclosing it in tick marks:

```python
a = True
```
... displays:

a = True"
the test is to eliminate the pages with the sales conditions.

**Reply by Larbo53** [reply]
environment :
IOS : BigSur 11.6
Python: 3.9.10
Pillow-9.0.1 
pdfminer.six-20211012

**Reply by jsvine** [reply]
>  How to knows the pdfplumber’s version?

Run `pip freeze` in the environment where you're using pdfplumber, and see what the line for `pdfplumber=...` says.

> 2 - I don't understand your second remark :

I was suggesting that you use Markdown code blocks, to make the code in your comments easier for others to read: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by Larbo53** [reply]
Hi,

with another invoice and the program, it's the same problem, with the same environment :
Result : 
{'CN00° 1fa6c0t3u0re04': [1], 'CN00° 1fa5c9t7u5re91': [2], 'CN00° 1fa5c9t3u3re47': [3], 'CN00° 1fa5c8t9u6re26': [4], 'CN00° 1fa5c8t5u7re32': [5], 'CN00° 1fa5c8t0u9re95': [6], 'CN00° 1fa5c7t2u9re65': [7]}
rec : (399.93, 41.83, 461.03, 58.03)
pdf join

What's wrong?

Sincerely
[Deroche_Factures C109046.pdf](https://github.com/jsvine/pdfplumber/files/8004166/Deroche_Factures.C109046.pdf)

**Reply by jsvine** [reply]
Thanks! I can reproduce your issue with this one, and now I see what's happening: The box you're drawing (displayed below in green) overlaps *slightly* with the bounding boxes of the characters (displayed in red) in the line above the one you want:

![image](https://user-images.githubusercontent.com/534702/152612289-2a6cf3af-7fca-4634-8d18-9a80e54dd8cd.png)

Because `.crop(...)` retains any object that *overlap* with the bounding box provided, those unwanted characters are still in the cropped version. There are a few different ways you could resolve this:

- Use `.within_bbox(...)`, which only retains objects *fulling within* the bounding box provided.
- Or: Change your bounding box to be slightly smaller, so that it doesn't overlap with the characters above.
- Or: Keep your current approach, but filter out any objects with a height less than, say `3`, which will remove the objects that only overlapped a little bit.

**Reply by Larbo53** [reply]
hi,

it's OK with .within_bbox(...)

Thanks a lot.

**Reply by Larbo53** [reply]
Good morning,

I'm still using the same process, but with a pdf file that contains several invoices and blank pages between some of them.
If the processing finds a blank page, it stops. What is the solution?

My code :
for i in range(0,len(doc.pages)):

    invoice=doc.pages[i].within_bbox(rec,relative = True).extract_text()
    if invoice[-2] not in ('0','1','2','3','4','5','6','7','8','9') and len(invoice)!=0:
        continue
    else:
Pages_Factures[doc.pages[i]]=invoice

with rec=(399.93, 41.83, 461.03, 58.03)

Thanks a lot
[deroche C108308.pdf](https://github.com/jsvine/pdfplumber/files/8010092/deroche.C108308.pdf)

**Comment by Larbo53** [comment]
Hi,

With pip3 list : the pdfplumber’s version is 0.6.0




Yves Larbodiere
EVARD
3, rue des Courtes Terres
95220 HERBLAY
portable : 07 81 08 41 00
mail : ***@***.***

**Reply by Larbo53** [reply]
Hi,

today, with another invoice with the same environnement and the same program, I have the same problem 👍 

metadata : {'Producer': 'PDFCreator Free 3.4.1', 'CreationDate': "D:20210701180908+02'00'", 'ModDate': "D:20210701180908+02'00'", 'Title': 'Facture C001603004 - Etat', 'Author': 'jmistretta', 'Subject': '', 'Keywords': '', 'Creator': 'PDFCreator Free 3.4.1'}

bbox : (399.93, 41.83, 461.03, 58.03)
return : {'CN00° 1fa6c0t3u0re04': [1], 'CN00° 1fa5c9t7u5re91': [2], 'CN00° 1fa5c9t3u3re47': [3], 'CN00° 1fa5c8t9u6re26': [4], 'CN00° 1fa5c8t5u7re32': [5], 'CN00° 1fa5c8t0u9re95': [6], 'CN00° 1fa5c7t2u9re65': [7]}
pdf joint

what's wrong?

Sincerely
[Deroche_Factures C109046.pdf](https://github.com/jsvine/pdfplumber/files/8004219/Deroche_Factures.C109046.pdf)

**Reply by Larbo53** [reply]
Hi,

I can't find my request of yesterday about the management of white pages in a pdf(attached) which groups several invoices.
here is my code :
for i in range(0,len(doc.pages)):

    invoice=doc.pages[i].within_bbox(rec,relative = True).extract_text()
    if invoice[-2] not in ('0','1','2','3','4','5','6','7','8','9') and len(invoice)!=0:
        continue
    else:
        Pages_Invoices[doc.pages[i]]=invoice

with rec=(399.93, 41.83, 461.03, 58.03)

The program is blocked by the first blank page. 
What is the solution?

Thank you.

Sincerely

Translated with www.DeepL.com/Translator (free version)
[deroche C108308.pdf](https://github.com/jsvine/pdfplumber/files/8010798/deroche.C108308.pdf)

**Comment by tegancp** [comment]
I would think you're getting an `IndexError` for a blank page, since your `invoice` variable in that case does not have an element at index -2 to access.  Try switching the order of your conditional, so that you're checking first that `invoice` has elements before accessing them.

**Reply by Larbo53** [reply]
hi,

The result is same.
What value returns an area or a blank page? By checking this value, the program could then read all the pages in the file.

Thanks for your feedback.
Best regards.

**Reply by Larbo53** [reply]
with 👍 

for i  in range(0,len(doc.pages)):#vérifier que c'est un num conforme c'est à dire qu'il contient au moins 2 chiffres

    invoice=doc.pages[i].within_bbox(rec,relative = True).extract_text()
    if invoice=='':
        continue

    elif invoice[-2] in ('0','1','2','3','4','5','6','7','8','9'):
        Pages_Factures[doc.pages[i]]=invoice

it's OK.

Thanks a lot.

Can you close this discussion?
